src/seraphis_main/contextual_enote_record_types.h

```diff
@@ -86,6 +86,79 @@ enum class SpEnoteSpentStatus : unsigned char
+    std::uint64_t enote_same_amount_ledger_index{static_cast<std::uint64_t>(-1)};
+    /// ledger index of the enote (-1 if index is unknown)
+    std::uint64_t enote_ledger_index{static_cast<std::uint64_t>(-1)};
+    /// origin status (off-chain by default)
+    SpEnoteOriginStatus origin_status{SpEnoteOriginStatus::OFFCHAIN};
+
+    /// associated memo field (none by default)
+    TxExtra memo{};
+};
+
+////
+// LegacyEnoteOriginContextV2
+// - info related to the transaction where a post-RingCT enote was found
+// - note that an enote may originate off-chain in a partial tx where the tx id is unknown
+///
+struct LegacyEnoteOriginContextV2 final
+{
+    /// block index of tx (-1 if index is unknown)
+    std::uint64_t block_index{static_cast<std::uint64_t>(-1)};
+    /// timestamp of tx's block (-1 if timestamp is unknown)
+    std::uint64_t block_timestamp{static_cast<std::uint64_t>(-1)};
+    /// tx id of the tx (0 if tx is unknown)
+    rct::key transaction_id{rct::zero()};
+    /// index of the enote in the tx's output set (-1 if index is unknown)
+    std::uint64_t enote_tx_index{static_cast<std::uint16_t>(-1)};
+    /// index of the enote with amount 0 in m_legacy_amount_counts (-1 if index is unknown)
+    std::uint64_t rct_enote_ledger_index{static_cast<std::uint64_t>(-1)};
+    /// ledger index of the enote (-1 if index is unknown)
+    std::uint64_t enote_ledger_index{static_cast<std::uint64_t>(-1)};
+    /// origin status (off-chain by default)
+    SpEnoteOriginStatus origin_status{SpEnoteOriginStatus::OFFCHAIN};
+
+    /// associated memo field (none by default)
+    TxExtra memo{};
+};
+
+////
+// LegacyEnoteOriginContextVariant
+// - variant of all legacy enote origin context types
+//
+// block_index_ref(): get the enote's block index
+// block_timestamp_ref(): get the enote's block timestamp
+// transaction_id_ref(): get the enote's transaction id
+// enote_ledger_index_ref(): get the enote's ledger index
+// origin_status_ref(): get the enote's origin status
+// enote_version_dependent_index_ref(): get the enote's
+//      enote_same_amount_ledger_index (V1) or rct_enote_ledger_index (V2)
+///
+using LegacyEnoteOriginContextVariant = tools::variant<LegacyEnoteOriginContextV1, LegacyEnoteOriginContextV2>;
+const std::uint64_t& block_index_ref(const LegacyEnoteOriginContextVariant &variant);
+const std::uint64_t& block_timestamp_ref(const LegacyEnoteOriginContextVariant &variant);
+const rct::key& transaction_id_ref(const LegacyEnoteOriginContextVariant &variant);
+const std::uint64_t& enote_ledger_index_ref(const LegacyEnoteOriginContextVariant &variant);
+const SpEnoteOriginStatus& origin_status_ref(const LegacyEnoteOriginContextVariant &variant);
+void origin_status_ref(const LegacyEnoteOriginContextVariant &variant, SpEnoteOriginStatus &origin_status_out);
+const std::uint64_t& enote_version_dependent_index_ref(const LegacyEnoteOriginContextVariant &variant);
```

Comment:
```suggestion
std::uint64_t block_index_ref(const LegacyEnoteOriginContextVariant &variant);
std::uint64_t block_timestamp_ref(const LegacyEnoteOriginContextVariant &variant);
const rct::key& transaction_id_ref(const LegacyEnoteOriginContextVariant &variant);
std::uint64_t enote_ledger_index_ref(const LegacyEnoteOriginContextVariant &variant);
SpEnoteOriginStatus origin_status_ref(const LegacyEnoteOriginContextVariant &variant);
std::uint64_t enote_version_dependent_index_ref(const LegacyEnoteOriginContextVariant &variant);
```

You only need an actual ref when it's not efficiently/trivially copyable. 16 bytes is a good rule of thumb as a cutoff.